autogen/agentchat/contrib/capabilities/transform_messages.py

```diff
@@ -0,0 +1,278 @@
+        """Applies token truncation to the conversation history.
+
+        Args:
+            messages (List[Dict]): The list of messages representing the conversation history.
+
+        Returns:
+            List[Dict]: A new list containing the truncated messages up to the specified token limits.
+        """
+        assert self._max_tokens_per_message is not None
+        assert self._max_tokens is not None
+
+        temp_messages = messages.copy()
+        processed_messages = []
+        processed_messages_tokens = 0
+
+        # calculate tokens for all messages
+        total_tokens = sum(_count_tokens(msg["content"]) for msg in temp_messages)
+
+        for msg in reversed(temp_messages):
+            msg["content"] = self._truncate_str_to_tokens(msg["content"])
+            msg_tokens = _count_tokens(msg["content"])
+
+            if processed_messages_tokens + msg_tokens > self._max_tokens:
+                break
+
+            # append the message to the beginning of the list to preserve order
+            processed_messages_tokens += msg_tokens
+            processed_messages.insert(0, msg)
+
+        if total_tokens > processed_messages_tokens:
+            print(
+                colored(
+                    f"Truncated {total_tokens - processed_messages_tokens} tokens. Tokens reduced from {total_tokens} to {processed_messages_tokens}",
+                    "yellow",
+                )
+            )
+
+        return processed_messages
+
+    def _truncate_str_to_tokens(self, contents: Union[str, List]) -> Union[str, List]:
+        if isinstance(contents, str):
+            return self._truncate_tokens(contents)
+        elif isinstance(contents, list):
+            return self._truncate_multimodal_text(contents)
+        else:
+            raise ValueError(f"Contents must be a string or a list of dictionaries. Received type: {type(contents)}")
+
+    def _truncate_multimodal_text(self, contents: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+        """Truncates text content within a list of multimodal elements, preserving the overall structure."""
+        tmp_contents = []
+        for content in contents:
+            if content["type"] == "text":
+                truncated_text = self._truncate_tokens(content["text"])
+                tmp_contents.append({"type": "text", "text": truncated_text})
+            else:
+                tmp_contents.append(content)
+        return tmp_contents
+
+    def _truncate_tokens(self, text: str) -> str:
+        encoding = tiktoken.encoding_for_model(self._model)  # Get the appropriate tokenizer
+
+        encoded_tokens = encoding.encode(text)
+        truncated_tokens = encoded_tokens[: self._max_tokens_per_message]
+        truncated_text = encoding.decode(truncated_tokens)  # Decode back to text
+
+        return truncated_text
+
+    def _validate_max_tokens(self, max_tokens: Optional[int] = None):
+        if max_tokens is not None and max_tokens < 0:
```

Comment:
Since we have `self._model`, should we also make sure the max_tokens is less than the model's capacity? 

```python
allowed_tokens = token_count_utils.get_max_token_limit(self._model)
if max_tokens > allowed_tokens:
    logger.warning(f"Max token was set to {max_tokens}, but {self._model} can only accept {allowed_tokens} tokens. Capping it to {allowed_tokens}.")
    max_tokens = allowed_tokens
```

Reply:
Very cool! Didn't know this existed. I'll add it as a validator.